File: PathFinding/RoutePlanner.cs
using PathFinding.AStar;

namespace PathFinding;

public static class RoutePlanner
{
    public static List<(ConnectionType, CorrectionType)> FindRoute(Node<(ConnectionType, CorrectionType)> start, Node<(ConnectionType, CorrectionType)> pickup, Node<(ConnectionType, CorrectionType)> drop)
    {
        var pickupRoute = GetRouteFromPath(AStar.AStar.FindPath(start, pickup, node => 0)); // TODO Implement heuristic
        var dropRoute = GetRouteFromPath(AStar.AStar.FindPath(pickup, drop, node => 0));
        dropRoute.RemoveAt(0); // Remove the first connection, which is the one we're currently on when we finish collecting the box

        var route = new List<(ConnectionType, CorrectionType)>(pickupRoute.Count + dropRoute.Count + 2);

        route.AddRange(pickupRoute);
        route.Add((ConnectionType.CollectBox, CorrectionType.None));
        route.AddRange(dropRoute);
        route.Add((ConnectionType.DropBox, CorrectionType.None));

        return route;
    }

    private static List<(ConnectionType, CorrectionType)> GetRouteFromPath(List<Node<(ConnectionType, CorrectionType)>> path)
    {
        return path.Zip(path.Skip(1))
                    .Select(nodes => nodes.First.Connections[nodes.Second].Item2)
                    .ToList();
    }
}

public enum ConnectionType
{
    Forward,
    Left,
    Right,
    BlueLine,
    CollectBox,
    DropBox
}

public enum CorrectionType
{
    Left,
    Right,
    None
}

File: Python/route_planner.py
from enum import Enum
from node import Node


class ConnectionType(Enum):
    Forward = 1
    Left = 2
    Right = 3
    BlueLine = 4
    CollectBox = 5
    DropBox = 6


class CorrectionType(Enum):
    Left = 1
    Right = 2
    NoCorrection = 3


def find_route(start: Node[tuple[ConnectionType, CorrectionType]], pickup: Node[tuple[ConnectionType, CorrectionType]],
               dropoff: Node[tuple[ConnectionType, CorrectionType]]) -> list[tuple[ConnectionType, CorrectionType]]:
    
    pickup_route: list = get_route_from_path(shortest_path(start, pickup))
    drop_route: list = get_route_from_path(shortest_path(pickup, dropoff))

    drop_route.pop(0)

    route = [
        *pickup_route,
        (ConnectionType.CollectBox, CorrectionType.NoCorrection),
        *drop_route,
        (ConnectionType.DropBox, CorrectionType.NoCorrection)
    ]

    return route

def get_route_from_path(nodes: list[Node[tuple[ConnectionType, CorrectionType]]]):
    return list(
        map(
            lambda nodes: nodes[0].connections[nodes[1]][1],
            zip(nodes, nodes[1:])
        )
    )